### PR TITLE
Python 3.6.8 - fix cross-compilation issue

### PR DESCRIPTION
**Pull request by roleoroleo**
_Motivation:_  I'm trying to use Home Assistant anda the last version deprecates Python 3.5
_Linked issues:_  https://github.com/SynoCommunity/spksrc/issues/3673

I cross-compiled it on x86_64 using a debian 9.9 with docker.
And I was able to test it on armada370.
I realized that the wheel files and the libraries contained still have the wrong notation "linux_x86_64" even if they are arm binaries.
Looking at the previous python 3.5.6 package I noticed that the situation is similar.
This is the output of the sys import executed on my DS214se:
```
roleo@DiskStation:~$ /var/packages/python3/target/bin/python3   -c 'import sqlite3; import _ctypes; import lxml; import sys; print ("OK:", sys.platform, sys.byteorder, "endian", sys.version)'
OK: linux little endian 3.6.8 (default, Jun 18 2019, 11:06:41)
[GCC 4.9.3 20150311 (prerelease)]
```

### CODE REVIEW

**Comment by effesseme** [comment]
I appreciate your work where I made an error, and I really appreciate your time to fix it.

In #3673 comment https://github.com/SynoCommunity/spksrc/issues/3673#issuecomment-500518542 we discussed how to check that the resulting cross-compiled binary works.

I think we're missing proof that the `ctypes` and some `cffi` can load correctly as well.

The key is whether the binary built on one arch runs correctly in the other.  On your `x86_64`-built python-3.6.8 for the `armada370`, please run the following, and please edit your top "is that OK?" to include this so that it's really obvious for the next iteration.

```
/var/packages/python3/target/bin/python3   -c 'import sqlite3; import _ctypes; import lxml; import sys; print ("OK:", sys.platform, sys.byteorder, "endian", sys.version)'
```

Specifically, I think the _ctypes exercising cffi, and I think sys exercising the _sysconfigdata, should show whether to get beta builds.  I would then suggest testing whether `home assistant` runs, since it uses a huge list of libraries brought in at postinstall (both wheel and pip-install, I think)

**Comment by chickenandpork** [comment]
(repost with correct ID)
I appreciate your work where I made an error, and I really appreciate your time to fix it.

In #3673 comment #3673 (comment) we discussed how to check that the resulting cross-compiled binary works.

I think we're missing proof that the ctypes and some cffi can load correctly as well.

The key is whether the binary built on one arch runs correctly in the other. On your x86_64-built python-3.6.8 for the armada370, please run the following, and please edit your top "is that OK?" to include this so that it's really obvious for the next iteration.

```
/var/packages/python3/target/bin/python3   -c 'import sqlite3; import _ctypes; import lxml; import sys; print ("OK:", sys.platform, sys.byteorder, "endian", sys.version)'
```

Specifically, I think the `_ctypes` exercising `cffi`, and I think `sys` exercising the `_sysconfigdata`, should show whether it's working despite any filename/path errors, and to get beta builds to test more broadly. I would then suggest testing whether `homeassistant` runs, since it uses a huge list of libraries brought in at postinstall (both wheel and pip-install, I think)

**Comment by roleoroleo** [comment]
I edited my previous post.
Regarding the errors on the names, I did several tests but I couldn't solve the issue.
The problem is related exclusively to the wheels created by "pip wheel".
I tried using the "--plat-name" build option but the compilation process does not complete successfully.

**Comment by chickenandpork** [comment]
> This is the output of the sys import executed on my DS214se:
To be clear, this is the output on your armada370 built on your x86_64?  I think it is, but it requires a hardware lookup rather than being immediately clear.

I think this is a successful sign though.  

Yet another check (platform.system) to try to detect issues:

```
/var/packages/python3/target/bin/python3   -c 'import sqlite3; import _ctypes; import platform; import lxml; import sys; print ("OK:", sys.platform, sys.byteorder, "endian", sys.version, "machine:", platform.machine())'
```

Does the "machine:" value look correct on your armada370 ?

My hidden agenda here is to be able to semi-automate later release validation.  Looking for a value on your armada370 HW that says something different than the `x86_64` on your build system

**Comment by roleoroleo** [comment]
```
roleo@DiskStation:~$ /var/packages/python3/target/bin/python3   -c 'import sqlite3; import _ctypes; import platform; import lxml; import sys; print ("OK:", sys.platform, sys.byteorder, "endian", sys.version, "machine:", platform.machine())'
OK: linux little endian 3.6.8 (default, Jun 18 2019, 11:06:41)
[GCC 4.9.3 20150311 (prerelease)] machine: armv7l
```

**Comment by chickenandpork** [comment]
Looks good to me then; reduces the customization/patch load by 56 lines or so.

Yves ?

**Comment by roleoroleo** [comment]
I'm working on another patch that also solves the problem of library and wheel names.

**Comment by schumi2004** [comment]
@roleoroleo 
Is this ready to be compiled?
I'm also trying to get HA up and running but old Pyhton packages fail on me and was hoping this would solve it ;)

Can't we just build an HA package maybe (if possible)?

**Comment by roleoroleo** [comment]
Unfortunately homeassistant does not work.
The bcrypt module does not compile, it fails during the pip wheel command.
If you remove bcrypt (just to try) the command completes successfully.

**Comment by schumi2004** [comment]
Yeah I also noticed that. Is bcrypt really needed then?

**Comment by roleoroleo** [comment]
I created a new patch to solve the bcrypt problem.
https://github.com/SynoCommunity/spksrc/pull/3731

**Comment by schumi2004** [comment]
@roleoroleo 
Still giving errors for me, sorry ;)
https://pastebin.com/YbQfHi5h

File: cross/python3/Makefile [hunk 1]
```diff
@@ -17,7 +17,8 @@ LICENSE  = PSF
 
 GNU_CONFIGURE = 1
 ADDITIONAL_CFLAGS = -D_LARGEFILE64_SOURCE -D_FILE_OFFSET_BITS=64 -L $(STAGING_INSTALL_PREFIX)/lib -I $(STAGING_INSTALL_PREFIX)/include
-CONFIGURE_ARGS  = --enable-shared --enable-ipv6 --without-ensurepip --enable-loadable-sqlite-extensions PYTHON_FOR_BUILD=$(HOSTPYTHON) PGEN_FOR_BUILD=$(HOSTPGEN)
+#CONFIGURE_ARGS  = --enable-shared --enable-ipv6 --without-ensurepip --enable-loadable-sqlite-extensions PYTHON_FOR_BUILD=$(HOSTPYTHON) PGEN_FOR_BUILD=$(HOSTPGEN)
```
Comment (ymartin59) [inline cross/python3/Makefile]: I would simply discard old line...

**Comment by ymartin59** [comment]
I am processing builds to confirm it works well... Then I would publish my own borgbackup (and maybe homeassistant) updates...

File: cross/python3/Makefile (hunk 1 above)
Comment (roleoroleo) [reply cross/python3/Makefile]: Sorry but I deleted all my forked repos and I don't known how to change the PR.

**Comment by ymartin59** [comment]
Repository and branch no longer available. Superseded by #3731

**Pull request closed without merging**